src/api/viewportFeaturesBinary.js

```diff
@@ -0,0 +1,133 @@
+import bboxPolygon from '@turf/bbox-polygon';
+import intersects from '@turf/boolean-intersects';
+import { prepareViewport, isTileFullVisible } from './viewportFeatures';
+
+const GEOMETRY_TYPES = Object.freeze({
+  Point: 0,
+  LineString: 1,
+  Polygon: 2
+});
+
+function addIntersectedFeaturesInTile({ map, data, viewportIntersection, type }) {
+  const indices = getIndices(data);
+  const { positions } = data;
+
+  for (let i = 0; i < indices.length - 1; i++) {
+    const startIndex = indices[i];
+    const endIndex = indices[i + 1];
+    const featureId = getFeatureId(data, startIndex);
+
+    if (!map.has(featureId)) {
+      const ringCoordinates = [];
+
+      for (let j = startIndex; j < endIndex; j++) {
+        ringCoordinates.push(
+          Array.from(
+            positions.value.subarray(j * positions.size, (j + 1) * positions.size)
+          )
+        );
+      }
+
+      if (intersects(getFeatureByType(ringCoordinates, type), viewportIntersection)) {
+        map.set(featureId, getPropertiesFromTile(data, startIndex));
+      }
+    }
```

Comment:
Could we extract this into a smaller method with a good name? Or at least the first part, dealing with coords. I think it deserves something like `const ringCoordinates = getRingCoordinatesFor(....whatever)`

Reply:
ok